cpp/src/arrow/util/bit_util.h

```diff
@@ -827,6 +835,167 @@ Status InvertBitmap(MemoryPool* pool, const uint8_t* bitmap, int64_t offset,
+      : buffer_(std::make_shared<Buffer>(data, BitUtil::BytesForBits(offset + length))),
+        offset_(offset),
+        length_(length) {}
+
+  Bitmap Slice(int64_t offset) const {
+    return Bitmap(buffer_, offset_ + offset, length_ - offset);
+  }
+
+  Bitmap Slice(int64_t offset, int64_t length) const {
+    return Bitmap(buffer_, offset_ + offset, length);
+  }
+
+  std::string ToString() const;
+
+  std::string Diff(const Bitmap& other) const;
+
+  bool GetBit(int64_t i) const { return BitUtil::GetBit(buffer_->data(), i + offset_); }
+
+  bool operator[](int64_t i) const { return GetBit(i); }
+
+  void SetBitTo(int64_t i, bool v) const {
+    BitUtil::SetBitTo(buffer_->mutable_data(), i + offset_, v);
+  }
+
+  /// \brief Visit bits of this bitmap as bitset<N>
+  template <size_t N, typename Visitor>
+  static void VisitBits(const Bitmap (&bitmaps)[N], Visitor&& visitor) {
+    auto bit_length = bitmaps[0].length();
+    std::bitset<N> bits;
+    for (int64_t bit_i = 0; bit_i < bit_length; ++bit_i) {
+      for (size_t i = 0; i < N; ++i) {
+        bits[i] = bitmaps[i].GetBit(bit_i);
+      }
+      visitor(bits);
+    }
+  }
+
+  /// \brief Visit bits of this bitmap as array<Word, N>
+  ///
+  /// If visiting by words, note that every Word containing a bit
+  /// must be accessible (and initialized, or Valgrind will barf).
+  /// This is satisfied by most Buffers in Arrow space (which are allocated with
+  /// maximum alignment and padded to contain whole cache lines).
+  ///
+  /// TODO(bkietz) allow for early termination
+  template <size_t N, typename Visitor,
+            typename Word =
+                typename internal::call_traits::argument_type<0, Visitor&&>::value_type>
+  static void VisitWords(const Bitmap (&bitmaps)[N], Visitor&& visitor) {
+    constexpr size_t kBitWidth = sizeof(Word) * 8;
+
+    View<Word> words[N];
+    int64_t offsets[N];
+    for (size_t i = 0; i < N; ++i) {
+      words[i] = bitmaps[i].template words<Word>();
+      offsets[i] = bitmaps[i].template word_offset<Word>();
+    }
+
+    std::array<Word, N> visited_words;
+
+    size_t word_count = 0;
+    for (auto word : words) {
+      // words[i].size() == words[j].size()    if offsets[i] == 0 and offsets[j] == 0
+      // words[i].size() == words[j].size()+1  if offsets[i] != 0 and offsets[j] == 0
+      word_count = std::max(word_count, word.size());
+    }
+
+    for (size_t word_i = 0; word_i < word_count - 1; ++word_i) {
+      for (size_t i = 0; i < N; ++i) {
+        if (offsets[i] == 0) {
+          visited_words[i] = words[i][word_i];
+        } else {
+          visited_words[i] = words[i][word_i] >> offsets[i];
+          visited_words[i] |= words[i][word_i + 1] << (kBitWidth - offsets[i]);
+        }
+      }
+      visitor(visited_words);
+    }
+
+    // handle last word
+    size_t word_i = word_count - 1;
+    for (size_t i = 0; i < N; ++i) {
+      if (word_i < words[i].size()) {
```

Comment:
What happens if not? Does the visitor get a stale value?

Reply:
Thanks for pointing this out; this was buggy and based on an erroneous assumption about offsets vs word count. I've removed that and added some comments which should clarify things